server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -434,6 +437,33 @@ public void copyFrom(Directory from, String src, IOContext context, ActionListen
         }
     }
 
+    /**
+     * Copies the segment from the remote directory to a local directory, preferably using an async mechanism using multi stream support.
+     * @param storeDirectory The directory where the segment needs to be copied to
+     * @param source The name of the source file
+     * @param destinationPath The file path for the directory where blob should be downloaded
+     * @param fileCompletionListener Async listener which should be notified once the segment is downloaded or in case of failure
+     */
+    public void copyTo(Directory storeDirectory, String source, Path destinationPath, ActionListener<String> fileCompletionListener) {
```

Comment:
The signature of this method doesn't really make sense to me. Here's a suggestion, though I still don't love it:

```java
/**
 * Copies an existing {@code source} file from this directory to a non-existent file (also
 * named {@code source}) in either {@code destinationDirectory} or {@code destinationPath}.
 * If the blob container backing this directory supports multipart downloads, the {@code source}
 * file will be downloaded (potentially in multiple concurrent parts) directly to
 * {@code destinationPath}. This method will return immediately and {@code fileCompletionListener}
 * will be notified upon completion.
 * <p>
 * If multipart downloads are not supported, then {@code source} file will be copied to a file named
 * {@code source} in a single part to {@code destinationDirectory}. The download will happen on the
 * calling thread and {@code fileCompletionListener} will be notified synchronously before this
 * method returns.
 *
 * @param source The source file name
 * @param destinationDirectory The destination directory (if multipart is not supported)
 * @param destinationPath The destination path (if multipart is supported)
 * @param fileCompletionListener The listener to notify of completion
 */
public void copyTo(String source, Directory destinationDirectory, Path destinationPath, ActionListener<String> fileCompletionListener) {

}
```

Reply:
It honestly seems simpler to do the check between the two modes (multipart to Path vs single part to directory) at the level where this is called, but the complication is that instance check and cast of an internal detail of RemoteSegmentStoreDirectory.

Reply:
I did think about it - given the experimental nature, I think it sits better here till it evolves since all the instances can call this method without having to worry about the implementation details. 
I'll update the signature to the suggested one - it reads better.